src/snapshot-editor/src/edit_memory.rs

```diff
@@ -0,0 +1,257 @@
+
+        // Diff file that has only data
+        let diff_data = rand::rand_bytes(50000);
+        diff_file.write_all(&diff_data).unwrap();
+        rebase(base_path, diff_path).unwrap();
+        check_file_content(base_file, &diff_data);
+    }
+
+    #[test]
+    fn test_rebase() {
+        // The filesystem punches holes only for blocks >= 4096.
+        // It doesn't make sense to test for smaller ones.
+        let block_sizes: &[usize] = &[4096, 8192];
+        for &block_size in block_sizes {
+            let mut expected_result = vec![];
+
+            let base = tempfile::TempFile::new().unwrap();
+            let diff = tempfile::TempFile::new().unwrap();
+
+            let mut base_file = base.as_file();
+            let mut diff_file = diff.as_file();
+
+            let base_path = base.as_path().to_path_buf();
+            let diff_path = diff.as_path().to_path_buf();
+
+            // 1. Populated block both in base and diff file
+            // block:     [ ]
+            // diff:      [ ]
+            // expected:  [d]
+            let base_block = rand::rand_bytes(block_size);
+            base_file.write_all(&base_block).unwrap();
+            let diff_block = rand::rand_bytes(block_size);
+            diff_file.write_all(&diff_block).unwrap();
+            expected_result.extend(diff_block);
+
+            // 2. Populated block in base file, hole in diff file
+            // block:     [ ] [ ]
+            // diff:      [ ] ___
+            // expected:  [d] [b]
+            let base_block = rand::rand_bytes(block_size);
+            base_file.write_all(&base_block).unwrap();
+            diff_file
+                .seek(SeekFrom::Current(block_size as i64))
+                .unwrap();
+            expected_result.extend(base_block);
+
+            // 3. Populated block in base file, zeroes block in diff file
+            // block:     [ ] [ ] [ ]
+            // diff:      [ ] ___ [0]
+            // expected:  [d] [b] [d]
+            let base_block = rand::rand_bytes(block_size);
+            base_file.write_all(&base_block).unwrap();
+            let diff_block = vec![0u8; block_size];
+            diff_file.write_all(&diff_block).unwrap();
+            expected_result.extend(diff_block);
+
+            // Rebase and check the result
+            rebase(base_path.clone(), diff_path.clone()).unwrap();
+            check_file_content(base_file, &expected_result);
+
+            // 4. The diff file is bigger
+            // block:     [ ] [ ] [ ]
+            // diff:      [ ] ___ [0] [ ]
+            // expected:  [d] [b] [d] [d]
+            let diff_block = rand::rand_bytes(block_size);
+            diff_file.write_all(&diff_block).unwrap();
+            expected_result.extend(diff_block);
+            // Rebase and check the result
+            rebase(base_path.clone(), diff_path.clone()).unwrap();
+            check_file_content(base_file, &expected_result);
+
+            // 5. The base file is bigger
```

Comment:
thanks for making the use cases in the test more clear with the visual comments!